File: packages/course-format/src/widgets.test.ts
import { describe, expect, it } from 'vitest';
import { validatePackage } from './validate';
import {
  WIDGET_FORBIDDEN_APIS,
  WIDGET_MAX_BYTES,
  WIDGET_MAX_LINE_BYTES,
  WIDGET_NAME_MAX,
  extractWidgetRefs,
} from './widgets';
import { codesOf, enc, withChapter } from './test-fixtures';

/**
 * The widget the task brief gives as the "this must validate clean" example:
 * a counter button, self-contained, nothing minified, nothing networked.
 * Reused across every test here that needs a widget with no complaints of its
 * own, so a failure in an unrelated rule's test can't be this snippet's fault.
 */
const VALID_WIDGET_HTML = `<style>button{font-size:2rem}</style>
<button id="b">0</button>
<script>
  let n = 0;
  document.getElementById('b').addEventListener('click', () => {
    n += 1;
    document.getElementById('b').textContent = String(n);
  });
</script>`;

describe('checkWidgets — tám luật', () => {
  it('WIDGET_TOO_LARGE: index.html của một widget vượt trần WIDGET_MAX_BYTES', () => {
    const files = withChapter('<div data-widget="big"></div>')
      .set('widgets/big/index.html', new Uint8Array(WIDGET_MAX_BYTES + 1));
    const findings = validatePackage(files).findings;
    const f = findings.find((x) => x.code === 'WIDGET_TOO_LARGE');
    expect(f?.path).toBe('widgets/big/index.html');
    expect(f?.detail).toContain(String(WIDGET_MAX_BYTES + 1));
  });

  it('WIDGET_TOO_LARGE: ĐÚNG trần thì sạch, trần + 1 thì báo', () => {
    const atCap = withChapter('<div data-widget="w"></div>')
      .set('widgets/w/index.html', enc('a'.repeat(WIDGET_MAX_BYTES)));
    expect(codesOf(atCap)).not.toContain('WIDGET_TOO_LARGE');
    const overCap = withChapter('<div data-widget="w"></div>')
      .set('widgets/w/index.html', enc('a'.repeat(WIDGET_MAX_BYTES + 1)));
    expect(codesOf(overCap)).toContain('WIDGET_TOO_LARGE');
  });

  it('WIDGET_LINE_TOO_LONG: một dòng vượt trần WIDGET_MAX_LINE_BYTES byte, detail nêu đúng số dòng', () => {
    const longLine = 'a'.repeat(WIDGET_MAX_LINE_BYTES + 1);
    const files = withChapter('<div data-widget="long"></div>')
      .set('widgets/long/index.html', enc(`<p>ok</p>\n${longLine}\n<p>ok nữa</p>`));
    const findings = validatePackage(files).findings;
    const f = findings.find((x) => x.code === 'WIDGET_LINE_TOO_LONG');
    expect(f?.path).toBe('widgets/long/index.html');
    // Dòng dài là dòng thứ 2 (1-based): dòng 1 là "<p>ok</p>", dòng 3 thì sạch.
    expect(f?.detail).toContain('dòng 2');
  });

  it('WIDGET_LINE_TOO_LONG: không báo khi mọi dòng đều dưới trần', () => {
    expect(codesOf(
      withChapter('<div data-widget="ok"></div>').set('widgets/ok/index.html', enc(VALID_WIDGET_HTML)),
    )).not.toContain('WIDGET_LINE_TOO_LONG');
  });

  it('WIDGET_LINE_TOO_LONG: ĐÚNG trần thì sạch, trần + 1 thì báo — tính theo BYTE, không theo ký tự', () => {
    const atCap = withChapter('<p>a</p>').set('widgets/w/index.html', enc('a'.repeat(WIDGET_MAX_LINE_BYTES)));
    expect(codesOf(atCap)).not.toContain('WIDGET_LINE_TOO_LONG');
    const overCap = withChapter('<p>a</p>').set('widgets/w/index.html', enc('a'.repeat(WIDGET_MAX_LINE_BYTES + 1)));
    expect(codesOf(overCap)).toContain('WIDGET_LINE_TOO_LONG');
    // Một dòng tiếng Việt NGẮN HƠN trần theo SỐ KÝ TỰ nhưng DÀI HƠN trần theo SỐ
    // BYTE ('ơ' chiếm 2 byte UTF-8) vẫn phải bị bắt — nếu luật lỡ đếm theo
    // `.length` (UTF-16 code unit) thay vì byte, ca này xanh oan.
    const vietnameseLine = 'ơ'.repeat(Math.floor(WIDGET_MAX_LINE_BYTES / 2) + 1);
    expect(vietnameseLine.length).toBeLessThan(WIDGET_MAX_LINE_BYTES);
    expect(codesOf(withChapter('<p>a</p>').set('widgets/w/index.html', enc(vietnameseLine)))).toContain('WIDGET_LINE_TOO_LONG');
  });

  it('WIDGET_BAD_NAME: tên có chữ hoa/gạch dưới bị bắt', () => {
    const files = withChapter('<p>a</p>').set('widgets/Bad_Name/index.html', enc(VALID_WIDGET_HTML));
    const findings = validatePackage(files).findings;
    const f = findings.find((x) => x.code === 'WIDGET_BAD_NAME');
    expect(f?.detail).toContain('Bad_Name');
  });

  it('WIDGET_BAD_NAME: tên dài quá WIDGET_NAME_MAX ký tự bị bắt, tên hợp lệ thì không', () => {
    const tooLong = 'a'.repeat(WIDGET_NAME_MAX + 1);
    expect(codesOf(
      withChapter('<p>a</p>').set(`widgets/${tooLong}/index.html`, enc(VALID_WIDGET_HTML)),
    )).toContain('WIDGET_BAD_NAME');
    const atCap = 'a'.repeat(WIDGET_NAME_MAX);
    expect(codesOf(
      withChapter('<p>a</p>').set(`widgets/${atCap}/index.html`, enc(VALID_WIDGET_HTML)),
    )).not.toContain('WIDGET_BAD_NAME');
  });

  it('WIDGET_FORBIDDEN_API: document.cookie/localStorage/sessionStorage/indexedDB đều bị bắt', () => {
    for (const api of WIDGET_FORBIDDEN_APIS) {
      const name = `w-${api.replace(/[^a-z]/gi, '').toLowerCase()}`;
      const files = withChapter(`<div data-widget="${name}"></div>`)
        .set(`widgets/${name}/index.html`, enc(`<script>${api}</script>`));
      const findings = validatePackage(files).findings;
      const f = findings.find((x) => x.code === 'WIDGET_FORBIDDEN_API');
      expect(f?.detail, `API "${api}" phải bị bắt`).toContain(api);
    }
  });

  it('WIDGET_FORBIDDEN_API: widget không đụng bốn API đó thì không bị báo', () => {
    expect(codesOf(
      withChapter('<div data-widget="ok"></div>').set('widgets/ok/index.html', enc(VALID_WIDGET_HTML)),
    )).not.toContain('WIDGET_FORBIDDEN_API');
  });

  it('WIDGET_EXTERNAL_URL: URL http(s):// bị bắt, kể cả nằm trong chú thích', () => {
    const files = withChapter('<div data-widget="net"></div>')
      .set('widgets/net/index.html', enc('<!-- https://evil.example/track.js -->'));
    const findings = validatePackage(files).findings;
    const f = findings.find((x) => x.code === 'WIDGET_EXTERNAL_URL');
    expect(f?.path).toBe('widgets/net/index.html');
    // Detail lấy nguyên văn từ task brief.
    expect(f?.detail).toBe('widget phải tự chứa: không tải gì từ mạng, kể cả trong chú thích — bỏ URL đi');
  });

  it('WIDGET_EXTRA_FILE: một tệp thứ hai cạnh index.html bị bắt, kể cả trong thư mục con', () => {
    const files = withChapter('<div data-widget="w"></div>')
      .set('widgets/w/index.html', enc(VALID_WIDGET_HTML))
      .set('widgets/w/assets/chart.js', enc('export function draw() {}'));
    const findings = validatePackage(files).findings;
    const f = findings.find((x) => x.code === 'WIDGET_EXTRA_FILE');
    expect(f?.path).toBe('widgets/w/assets/chart.js');
  });

  it('WIDGET_MISSING: chương tham chiếu widget không có trong gói, path là chương chứa ref', () => {
    const files = withChapter('<div data-widget="khong-co"></div>');
    const findings = validatePackage(files).findings;
    const f = findings.find((x) => x.code === 'WIDGET_MISSING');
    expect(f?.path).toBe('chapters/c1.html');
    expect(f?.detail).toContain('khong-co');
  });

  it('WIDGET_ORPHAN: widget có thật nhưng không chương nào tham chiếu, path là widgets/<tên>/index.html', () => {
    const files = withChapter('<p>không nhắc tới widget nào</p>')
      .set('widgets/mo-coi/index.html', enc(VALID_WIDGET_HTML));
    const findings = validatePackage(files).findings;
    const f = findings.find((x) => x.code === 'WIDGET_ORPHAN');
    expect(f?.path).toBe('widgets/mo-coi/index.html');
  });

  // Fix round 1, finding 1: một thư mục widgets/<tên>/ CHỈ có tệp thừa (không
  // có index.html) không phải một widget THẬT — dù `byName.has(name)` vẫn
  // đúng cho nó. Trước bản sửa này, chương tham chiếu một thư mục như vậy
  // không hề bị WIDGET_MISSING (thư mục "có tồn tại"), còn nếu không được
  // tham chiếu thì lại bị WIDGET_ORPHAN trỏ tới một index.html chưa từng
  // được tạo ra. Cả hai đều là báo sai.
  it('WIDGET_MISSING: thư mục widget CHỈ có tệp thừa, không có index.html, mà chương tham chiếu tới — vẫn phải báo thiếu', () => {
    const files = withChapter('<div data-widget="foo"></div>')
      .set('widgets/foo/notes.txt', enc('không phải index.html'));
    const findings = validatePackage(files).findings;
    expect(findings.map((f) => f.code)).toContain('WIDGET_MISSING');
    expect(findings.map((f) => f.code)).toContain('WIDGET_EXTRA_FILE'); // notes.txt vẫn là tệp thừa, độc lập với WIDGET_MISSING
    const missing = findings.find((f) => f.code === 'WIDGET_MISSING');
    expect(missing?.path).toBe('chapters/c1.html');
  });

  it('WIDGET_ORPHAN: thư mục widget CHỈ có tệp thừa, không có index.html, không được tham chiếu — KHÔNG được báo mồ côi', () => {
    const files = withChapter('<p>không nhắc tới widget nào</p>')
      .set('widgets/foo/notes.txt', enc('không phải index.html'));
    const findings = validatePackage(files).findings;
    // Không có index.html thì không có gì để "mồ côi" — path widgets/foo/index.html
    // chưa từng tồn tại trong gói. WIDGET_EXTRA_FILE là đủ để chỉ ra vấn đề thật.
    expect(findings.map((f) => f.code)).not.toContain('WIDGET_ORPHAN');
    expect(findings.map((f) => f.code)).toContain('WIDGET_EXTRA_FILE');
  });

  it('WIDGET_MISSING: data-widget="" không mang tên — detail phải nói đó là ô trống, không phải bịa ra một tên widget rỗng', () => {
    const files = withChapter('<div data-widget=""></div>');
    const findings = validatePackage(files).findings;
    const f = findings.find((x) => x.code === 'WIDGET_MISSING');
    expect(f?.path).toBe('chapters/c1.html');
    // Không được đọc như một cái tên widget hợp lệ tên là "" — câu phải nói
    // đây là một ô/placeholder chưa được điền tên, không phải trỏ tới
    // "widgets//index.html".
    expect(f?.detail).not.toContain('widgets//index.html');
    expect(f?.detail).toMatch(/không mang tên|chưa.*tên|placeholder/i);
  });
});

describe('extractWidgetRefs', () => {
  it('đọc được data-widget dù thuộc tính viết hoa/lẫn lộn (DATA-WIDGET)', () => {
    // parse5's tokenizer ASCII-lowercases every attribute NAME as part of
    // tokenizing (same fact `validate.ts`'s EVENT_HANDLER_NAME_RE comment
    // relies on for ONERROR) — so this is really a test that extractWidgetRefs
    // goes through that tokenizer rather than a hand-rolled, case-sensitive
    // string search.
    expect(extractWidgetRefs('<div DATA-WIDGET="demo"></div>')).toEqual(['demo']);
    expect(extractWidgetRefs('<div Data-Widget="demo2"></div>')).toEqual(['demo2']);
  });
});

it('gói có widget hợp lệ + chương trỏ đúng → 0 finding', () => {
  const files = withChapter('<p>Trước đó.</p><div data-widget="dem"></div><p>Sau đó.</p>')
    .set('widgets/dem/index.html', enc(VALID_WIDGET_HTML));
  expect(validatePackage(files)).toEqual({ ok: true, findings: [] });
});

File: apps/api/internal/pkgcheck/widgets_test.go
// Unit tests for the eight widget rules (widgets.go) beyond what
// fixtures/format-v2/hostile/ covers via contract_test.go's
// TestHostileWidgetCases.
//
// Three rules need GENERATED content the corpus deliberately does not
// carry in git — a 128 KiB widget, a 500-byte line, an over-length name —
// mirroring how TestFormTagIsFlagged/TestTagAttrFloodIsFlagged in
// pkgcheck_test.go cover FORM_TAG/TAG_ATTR_FLOOD, the two content rules
// with no hostile/<case> directory of their own either.
//
// The rest of this file pins the "details that bite" the task brief calls
// out by name: isRealWidget's directory-with-only-a-stray-file distinction
// (widgets.ts's own review round 1 finding), WIDGET_MISSING's per-chapter
// dedup, an empty data-widget="" getting its own detail sentence instead of
// naming a widget called "", data-widget read case-insensitively, and the
// path-sorted (not name-sorted) order groupWidgetFiles produces.
package pkgcheck

import (
	"fmt"
	"strings"
	"testing"
)

// validWidgetHTML is the task brief's own "must validate clean" example: a
// self-contained counter button, nothing minified, nothing networked.
// Matches packages/course-format/src/widgets.test.ts's VALID_WIDGET_HTML so
// a failure in one rule's test can't be blamed on an unrelated one.
const validWidgetHTML = `<style>button{font-size:2rem}</style>
<button id="b">0</button>
<script>
  let n = 0;
  document.getElementById('b').addEventListener('click', () => {
    n += 1;
    document.getElementById('b').textContent = String(n);
  });
</script>`

// mustValidate runs Validate over files (built with buildZip) and fails the
// test on any unexpected system error — every test below is exercising a
// hostile PACKAGE, never a Go-level failure, so a non-nil err is always a
// test-harness bug, not the thing under test.
func mustValidate(t *testing.T, files map[string]string) []Finding {
	t.Helper()
	findings, _, err := Validate(buildZip(t, files))
	if err != nil {
		t.Fatalf("unexpected system error: %v", err)
	}
	return findings
}

func findByCode(findings []Finding, code string) *Finding {
	for i := range findings {
		if findings[i].Code == code {
			return &findings[i]
		}
	}
	return nil
}

// --- WIDGET_TOO_LARGE: generated content, no fixture --------------------

func TestWidgetTooLargeGenerated(t *testing.T) {
	atCap := mustValidate(t, map[string]string{
		"manifest.json":        minimalManifest,
		"chapters/c1.html":     `<div data-widget="w"></div>`,
		"widgets/w/index.html": strings.Repeat("a", widgetMaxBytes),
	})
	if hasCode(atCap, "WIDGET_TOO_LARGE") {
		t.Fatalf("exactly widgetMaxBytes must not trip WIDGET_TOO_LARGE, got %+v", atCap)
	}

	overCap := mustValidate(t, map[string]string{
		"manifest.json":        minimalManifest,
		"chapters/c1.html":     `<div data-widget="w"></div>`,
		"widgets/w/index.html": strings.Repeat("a", widgetMaxBytes+1),
	})
	f := findByCode(overCap, "WIDGET_TOO_LARGE")
	if f == nil {
		t.Fatalf("widgetMaxBytes+1 must trip WIDGET_TOO_LARGE, got %+v", overCap)
	}
	if f.Path != "widgets/w/index.html" {
		t.Errorf("WIDGET_TOO_LARGE path = %q, want widgets/w/index.html", f.Path)
	}
	if !strings.Contains(f.Detail, fmt.Sprintf("%d", widgetMaxBytes+1)) {
		t.Errorf("detail should name the actual byte count %d, got %q", widgetMaxBytes+1, f.Detail)
	}
}

// --- WIDGET_LINE_TOO_LONG: generated content, no fixture -----------------

func TestWidgetLineTooLongGenerated(t *testing.T) {
	atCap := mustValidate(t, map[string]string{
		"manifest.json":        minimalManifest,
		"chapters/c1.html":     `<p>a</p>`,
		"widgets/w/index.html": strings.Repeat("a", widgetMaxLineBytes),
	})
	if hasCode(atCap, "WIDGET_LINE_TOO_LONG") {
		t.Fatalf("a line at exactly widgetMaxLineBytes must not trip WIDGET_LINE_TOO_LONG, got %+v", atCap)
	}

	longLine := strings.Repeat("a", widgetMaxLineBytes+1)
	overCap := mustValidate(t, map[string]string{
		"manifest.json":           minimalManifest,
		"chapters/c1.html":        `<div data-widget="long"></div>`,
		"widgets/long/index.html": "<p>ok</p>\n" + longLine + "\n<p>ok too</p>",
	})
	f := findByCode(overCap, "WIDGET_LINE_TOO_LONG")
	if f == nil {
		t.Fatalf("widgetMaxLineBytes+1 must trip WIDGET_LINE_TOO_LONG, got %+v", overCap)
	}
	// The long line is line 2 (1-based): line 1 is "<p>ok</p>".
	if !strings.Contains(f.Detail, "line 2") {
		t.Errorf("detail should name line 2, got %q", f.Detail)
	}

	// A Vietnamese line SHORTER than the cap in RUNE COUNT but LONGER in
	// BYTE COUNT ('ơ' is 2 bytes in UTF-8) must still be caught — pins the
	// byte-not-character measurement the task brief calls out explicitly,
	// mirroring widgets.test.ts's own pinned diacritics test.
	vnLine := strings.Repeat("ơ", widgetMaxLineBytes/2+1)
	if len([]rune(vnLine)) >= widgetMaxLineBytes {
		t.Fatalf("test fixture bug: vnLine has %d runes, must be under widgetMaxLineBytes to prove the byte-not-rune point", len([]rune(vnLine)))
	}
	if len(vnLine) <= widgetMaxLineBytes {
		t.Fatalf("test fixture bug: vnLine has %d bytes, must be OVER widgetMaxLineBytes to prove the byte-not-rune point", len(vnLine))
	}
	vnFindings := mustValidate(t, map[string]string{
		"manifest.json":        minimalManifest,
		"chapters/c1.html":     `<p>a</p>`,
		"widgets/w/index.html": vnLine,
	})
	if !hasCode(vnFindings, "WIDGET_LINE_TOO_LONG") {
		t.Fatalf("a Vietnamese line shorter in RUNES but longer in BYTES than widgetMaxLineBytes must still trip WIDGET_LINE_TOO_LONG — got %+v", vnFindings)
	}
}

// --- WIDGET_BAD_NAME: generated content, no fixture -----------------------

func TestWidgetBadNameGenerated(t *testing.T) {
	tooLong := strings.Repeat("a", widgetNameMax+1)
	findings := mustValidate(t, map[string]string{
		"manifest.json":                      minimalManifest,
		"chapters/c1.html":                   `<p>a</p>`,
		"widgets/" + tooLong + "/index.html": validWidgetHTML,
	})
	if !hasCode(findings, "WIDGET_BAD_NAME") {
		t.Fatalf("a name over widgetNameMax characters must trip WIDGET_BAD_NAME, got %+v", findings)
	}

	atCapName := strings.Repeat("a", widgetNameMax)
	findings = mustValidate(t, map[string]string{
		"manifest.json":                        minimalManifest,
		"chapters/c1.html":                     fmt.Sprintf(`<div data-widget="%s"></div>`, atCapName),
		"widgets/" + atCapName + "/index.html": validWidgetHTML,
	})
	if hasCode(findings, "WIDGET_BAD_NAME") {
		t.Fatalf("a name at exactly widgetNameMax characters must not trip WIDGET_BAD_NAME, got %+v", findings)
	}

	findings = mustValidate(t, map[string]string{
		"manifest.json":               minimalManifest,
		"chapters/c1.html":            `<p>a</p>`,
		"widgets/Bad_Name/index.html": validWidgetHTML,
	})
	f := findByCode(findings, "WIDGET_BAD_NAME")
	if f == nil {
		t.Fatalf("uppercase/underscore in a widget name must trip WIDGET_BAD_NAME, got %+v", findings)
	}
	if !strings.Contains(f.Detail, "Bad_Name") {
		t.Errorf("detail should name the offending name, got %q", f.Detail)
	}
}

// --- WIDGET_FORBIDDEN_API: all four APIs, not only the one the fixture --
// --- corpus happens to use (localStorage) ------------------------------

func TestWidgetForbiddenAPIAllFourCaught(t *testing.T) {
	for _, api := range widgetForbiddenAPIs {
		api := api
		t.Run(api, func(t *testing.T) {
			name := "w-" + sanitizeForWidgetName(api)
			findings := mustValidate(t, map[string]string{
				"manifest.json":                   minimalManifest,
				"chapters/c1.html":                fmt.Sprintf(`<div data-widget="%s"></div>`, name),
				"widgets/" + name + "/index.html": fmt.Sprintf("<script>%s</script>", api),
			})
			f := findByCode(findings, "WIDGET_FORBIDDEN_API")
			if f == nil {
				t.Fatalf("API %q must be caught, got %+v", api, findings)
			}
			if !strings.Contains(f.Detail, api) {
				t.Errorf("detail should name the API %q, got %q", api, f.Detail)
			}
		})
	}
}

func sanitizeForWidgetName(s string) string {
	var b strings.Builder
	for _, r := range strings.ToLower(s) {
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// --- WIDGET_EXTRA_FILE: even nested in a subdirectory --------------------

func TestWidgetExtraFileCaughtEvenInSubdirectory(t *testing.T) {
	findings := mustValidate(t, map[string]string{
		"manifest.json":             minimalManifest,
		"chapters/c1.html":          `<div data-widget="w"></div>`,
		"widgets/w/index.html":      validWidgetHTML,
		"widgets/w/assets/chart.js": `export function draw() {}`,
	})
	f := findByCode(findings, "WIDGET_EXTRA_FILE")
	if f == nil {
		t.Fatalf("expected WIDGET_EXTRA_FILE, got %+v", findings)
	}
	if f.Path != "widgets/w/assets/chart.js" {
		t.Errorf("WIDGET_EXTRA_FILE path = %q, want widgets/w/assets/chart.js", f.Path)
	}
}

// --- isRealWidget: a directory with ONLY a stray file is not a widget ---
// widgets.ts's own review round 1 finding, both directions.

func TestWidgetDirWithOnlyStrayFileIsNotARealWidget(t *testing.T) {
	t.Run("referenced by a chapter still reports WIDGET_MISSING", func(t *testing.T) {
		findings := mustValidate(t, map[string]string{
			"manifest.json":         minimalManifest,
			"chapters/c1.html":      `<div data-widget="foo"></div>`,
			"widgets/foo/notes.txt": "not an index.html",
		})
		missing := findByCode(findings, "WIDGET_MISSING")
		if missing == nil {
			t.Fatalf("a widgets/foo/ directory with no index.html is not a real widget — a chapter referencing it must still get WIDGET_MISSING, got %+v", findings)
		}
		if missing.Path != "chapters/c1.html" {
			t.Errorf("WIDGET_MISSING path = %q, want chapters/c1.html", missing.Path)
		}
		if !hasCode(findings, "WIDGET_EXTRA_FILE") {
			t.Errorf("notes.txt is still a stray file, independent of WIDGET_MISSING — got %+v", findings)
		}
	})

	t.Run("unreferenced does NOT report WIDGET_ORPHAN", func(t *testing.T) {
		findings := mustValidate(t, map[string]string{
			"manifest.json":         minimalManifest,
			"chapters/c1.html":      `<p>no widget mentioned here</p>`,
			"widgets/foo/notes.txt": "not an index.html",
		})
		if hasCode(findings, "WIDGET_ORPHAN") {
			t.Fatalf("widgets/foo/index.html was never created — there is nothing to be \"orphaned\"; WIDGET_EXTRA_FILE alone is the right diagnosis, got %+v", findings)
		}
		if !hasCode(findings, "WIDGET_EXTRA_FILE") {
			t.Fatalf("expected WIDGET_EXTRA_FILE, got %+v", findings)
		}
	})
}

// --- WIDGET_MISSING: per-chapter dedup, and the empty-name placeholder --

func TestWidgetMissingDedupedPerChapterReference(t *testing.T) {
	findings := mustValidate(t, map[string]string{
		"manifest.json":    minimalManifest,
		"chapters/c1.html": `<div data-widget="khong-co"></div><div data-widget="khong-co"></div>`,
	})
	count := 0
	for _, f := range findings {
		if f.Code == "WIDGET_MISSING" {
			count++
		}
	}
	if count != 1 {
		t.Fatalf("a chapter referencing the same missing widget twice must report WIDGET_MISSING once, not %d — got %+v", count, findings)
	}
}

func TestWidgetMissingEmptyDataWidgetGetsPlaceholderDetail(t *testing.T) {
	findings := mustValidate(t, map[string]string{
		"manifest.json":    minimalManifest,
		"chapters/c1.html": `<div data-widget=""></div>`,
	})
	f := findByCode(findings, "WIDGET_MISSING")
	if f == nil {
		t.Fatalf("expected WIDGET_MISSING, got %+v", findings)
	}
	// Must not read like a widget literally named "" living at
	// widgets//index.html — that is not a diagnosis of the chapter's
	// actual mistake (an unfinished placeholder), it just looks broken.
	if strings.Contains(f.Detail, "widgets//index.html") {
		t.Fatalf("an empty data-widget must not be reported as a reference to widgets//index.html — got detail %q", f.Detail)
	}
}

// --- data-widget is read case-insensitively, via content.go's tokenizer -

func TestWidgetMissingReadsDataWidgetCaseInsensitively(t *testing.T) {
	findings := mustValidate(t, map[string]string{
		"manifest.json":    minimalManifest,
		"chapters/c1.html": `<div DATA-WIDGET="khong-co"></div>`,
	})
	if !hasCode(findings, "WIDGET_MISSING") {
		t.Fatalf("DATA-WIDGET (uppercase attribute name) must be read the same as data-widget — HTML attribute names are case-insensitive, got %+v", findings)
	}
}

// --- WIDGET_ORPHAN: referenced by either of two chapters is not orphan --

const twoChapterWidgetManifest = `{
  "id": "unit-test",
  "title": "Unit test package",
  "description": "",
  "lang": "vi",
  "version": "1.0.0",
  "runtime": "^1",
  "license": "CC0-1.0",
  "authors": [{"name": "test"}],
  "generatedBy": "ai",
  "parts": [{"title": "Part", "chapters": [
    {"id": "c1", "num": "1", "title": "C1", "short": "C1", "file": "chapters/c1.html"},
    {"id": "c2", "num": "2", "title": "C2", "short": "C2", "file": "chapters/c2.html"}
  ]}]
}`

func TestWidgetReferencedByEitherOfTwoChaptersIsNotOrphan(t *testing.T) {
	findings := mustValidate(t, map[string]string{
		"manifest.json":             twoChapterWidgetManifest,
		"chapters/c1.html":          `<p>no widget here</p>`,
		"chapters/c2.html":          `<div data-widget="shared"></div>`,
		"widgets/shared/index.html": validWidgetHTML,
	})
	if hasCode(findings, "WIDGET_ORPHAN") {
		t.Fatalf("a widget referenced by ONE of two chapters must not be reported orphan, got %+v", findings)
	}
	if len(findings) != 0 {
		t.Fatalf("expected zero findings, got %+v", findings)
	}
}

// --- Ordering: groupWidgetFiles preserves PATH-sorted order, not --------
// --- alphabetical NAME order ---------------------------------------------

// TestWidgetFindingOrderFollowsSortedPathsNotSortedNames pins the subtle
// fact documented on groupWidgetFiles: "widgets/Big-Two/" sorts BEFORE
// "widgets/Big/" as byte strings ('-' is 0x2D, '/' is 0x2F), even though
// "Big" sorts before "Big-Two" as bare strings. If groupWidgetFiles were
// changed to sort widget names directly instead of preserving path-walk
// order, this test would catch the regression by seeing the two
// WIDGET_BAD_NAME findings swap places.
func TestWidgetFindingOrderFollowsSortedPathsNotSortedNames(t *testing.T) {
	findings := mustValidate(t, map[string]string{
		"manifest.json":              minimalManifest,
		"chapters/c1.html":           `<p>a</p>`,
		"widgets/Big/index.html":     validWidgetHTML,
		"widgets/Big-Two/index.html": validWidgetHTML,
	})
	var order []string
	for _, f := range findings {
		if f.Code == "WIDGET_BAD_NAME" {
			order = append(order, f.Path)
		}
	}
	want := []string{"widgets/Big-Two", "widgets/Big"}
	if len(order) != 2 || order[0] != want[0] || order[1] != want[1] {
		t.Fatalf("WIDGET_BAD_NAME order = %v, want %v (path-sorted first-seen order, not name-sorted)", order, want)
	}
}

// --- Zero findings for a correctly-shipped widget -------------------------

func TestPackageWithValidWidgetAndCorrectReferenceHasZeroFindings(t *testing.T) {
	findings := mustValidate(t, map[string]string{
		"manifest.json":          minimalManifest,
		"chapters/c1.html":       `<p>Before.</p><div data-widget="dem"></div><p>After.</p>`,
		"widgets/dem/index.html": validWidgetHTML,
	})
	if len(findings) != 0 {
		t.Fatalf("expected zero findings for a valid widget correctly referenced, got %+v", findings)
	}
}
